Remove ValueFetcher depedendency from MapperService

javanna opened the pull request:
ValueFetcher needs only one method from MapperService: sourcePaths. The signature of MappedFieldType#valueFetcher requires MapperService as an argument which is unfortunate as that is one of the reasons why FetchContext exposes the whole MapperService.

This can be easily removed by having a specific lightweight object to carry around instead of the MapperService, that exposes only the needed method.

This change is marked `breaking-java` because it changes the signature of the `MappedFieldType#valueFetcher` method to take `QueryShardContext` as argument instead of `MapperService`. `MapperPlugin`s that plug in a custom field type implementation will have to adapt their code to make it compile with Elasticsearch 7.11 and following versions.

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

romseygeek reviewed (approved):
LGTM, very nice.

jimczi commented on server/src/main/java/org/elasticsearch/index/mapper/MappedFieldType.java:
```diff
@@ -105,7 +105,7 @@ public MappedFieldType(String name, boolean isIndexed, boolean isStored,
      * could cause a search retrieving multiple fields (like "fields": ["*"]) to fail.
      */
-    public abstract ValueFetcher valueFetcher(MapperService mapperService, SearchLookup searchLookup, @Nullable String format);
+    public abstract ValueFetcher valueFetcher(ValueFetcher.Context valueFetcherContext, SearchLookup searchLookup, @Nullable String format);
```
Why not using `QueryShardContext` ? We could also remove the `SearchLookup` argument.

nik9000 replied:
> Why not using `QueryShardContext` ? We could also remove the `SearchLookup` argument.

I'd prefer not to use `QueryShardContext` everywhere. It really is "too big" as it is.

For what it is worth, we could totally add `SearchLookup` to `ValueFetcher.Context` if we wanted.

But I've made three attempts at this change so far and we didn't like them so I'm kind of done having opinions here. Do whatever makes folks happy.

javanna replied:
QueryShardContext is a huge object that I don't think we want to spread around more than it already is. While we remove the dependency from MapperService, I think that we should try and be careful with encapsulation and who gets access to what. `sourcePath` is currently not exposed anywhere other than `MapperService`, and we will have to find a way to expose it so we can progress with removing direct MapperService usages, but we want to be careful with adding new methods to all of the existing contexts, that already have many methods.

javanna replied:
I updated and added QueryShardContext as an argument. I am not sure about SearchLookup because it needs to retrieved specifically through QueryShardContext#newFetchLookup and reused. Looks like we should improve this as its confusing that you can get the lookup also from the context, though it would be the wrong one. I will leave this for another time though.

jimczi replied:
+1 to solve this as a follow up.

jimczi reviewed (commented):
I left one comment

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/ValueFetcher.java:
```diff
@@ -50,4 +51,8 @@
      */
     default void setNextReader(LeafReaderContext context) {}
+
+    interface Context {
```
I'm not sure that this is better than my last attempt. I think it's up to @jtibshirani to make the call on this one.

javanna replied:
I think that the difference is subtle, but I believe this should address the concerns around tracking usages. At the same time we are introducing a lightweight abstraction that clearly encapsulates what is needed to fetch values, which is much harder to read when functions are carried around. I would love for Julie to have a look and let us know what she thinks though.

jimczi commented:
I think it's important to be consistent here. IMO the `QueryShardContext` is the **lightweight** object that we use when parsing/building/fetching in the query or the fetch phase so it's logic to expose it entirely. Look at the existing query  methods  in `MappedFieldType`, they already expose the `QueryShardContext`. We could argue that they don't need the entire context but that would make the API more complex to  use. 
My point here is that the `QSC`  is the right abstraction to expose for consistency.  We can ensure that nothing unsafe leaks in a central point.

javanna committed:
Remove ValueFetcher depedendency from MapperService

The signature of MappedFieldType#valueFetcher requires MapperService as an argument which is unfortunate as that is one of the reasons why FetchContext exposes the whole MapperService.

Such use of MapperService can be replaced with exposing the QueryShardContext which encapsulates the MapperService.

javanna committed:
fix test

romseygeek commented on modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/RankFeatureFieldMapper.java:
```diff
@@ -114,11 +114,11 @@ public Query existsQuery(QueryShardContext context) {
 
         @Override
-        public ValueFetcher valueFetcher(MapperService mapperService, SearchLookup searchLookup, String format) {
+        public ValueFetcher valueFetcher(QueryShardContext context, SearchLookup searchLookup, String format) {
```
If we're passing QueryShardContext I think that means we can avoid passing SearchLookup here?

javanna replied:
see my comment above, I don't think so, because we need to pass through the fetch lookup that is created in FetchContext?

romseygeek commented on server/src/main/java/org/elasticsearch/index/query/QueryShardContext.java:
```diff
@@ -270,6 +270,14 @@ public ObjectMapper getObjectMapper(String name) {
+        return mapperService.isMetadataField(field);
+    }
+
+    public Set<String> sourcePath(String fullName) {
```
One nit: maybe this should be `sourcePaths`, given that it can return more than one?

javanna replied:
yes, in my iterations I did both, I went for singular for consistency with the name of the method in MapperService. I don't know what is more important between using the same name or using the correct name here. Maybe we should rename both as a followup?

romseygeek replied:
+1 for a followup

romseygeek reviewed (approved):
One nit, but LGTM otherwise.

jtibshirani reviewed (commented):
Sorry for the slow review, I like the approach taken in the PR to pass in `QueryShardContext`. +1 to think about a follow-up refactor where we avoid passing around `QueryShardContext` and `SearchLookup` separately.
